### ISSUE TITLE
Another week, another package

### ISSUE THREAD

**Issue by LTLA** [LTLA/GenomicInteractions#1]
Technically depends on Bioconductor/S4Vectors#36 for some of the tests, but otherwise it `check`s out. Tagging @liz-is, @mgperry for thoughts in relation to ComputationalRegulatoryGenomicsICL/GenomicInteractions#37.

**Comment by LTLA** [comment]
Refactoring is done for `GRangesFactor` usage; this can now be considered ready to go.

**Comment by liz-is** [comment]
Hey, I got a chance to look at this today. Looks like it's still waiting on some of the `GRangesFactor` overlaps functions to be accepted into `GenomicRanges`? I had to install your fork of `GenomicRanges`. Anyway, I had a look through the code and vignettes - looks very streamlined and generally looks good to me. 

There's one issue I kept running into:
```
order(combined)
Error: C stack usage  7969256 is too close to the limit
```

This happens with a bunch of other functions but seems to trace back to calls to `order()` and within that to `xtfrm()` (I think). Anything related to C is beyond my experience so I didn't look into this further.

A few minor things I noticed:

* It would be nice to have an option to drop the regions/levels that are not in use when calling `anchors(gi, type =1)` or `regions()`, analogous to dropping levels in a factor.

* The `class.Rmd` vignette says that annotation should propagate to anchors retrieved with `anchors()`, but it doesn't. 

* The `utilities.Rmd` vignette says that `pairdist(gi, type = "diagonal")` should work, but it only works with "diag". 

* I also found the description of this type a bit confusing, specifically, it's not clear to me what you mean by "this corresponds to a diagonal on the interaction space". (Are the other distances not *also* along a diagonal in the interaction space?) For me it would be more intuitive to phrase this in terms of the number of bins between regions (when regions are genomic bins).

* I noticed that `pairdist()` returns absolute numbers for all types except "diag". Do you have a use case in mind that needs relative rather than absolute distances in bins? Otherwise, IMO it would be preferable for this behaviour to be the same across the different types.

**Comment by LTLA** [comment]
> ```
> Error: C stack usage  7969256 is too close to the limit
> ```

Yes, I think this should have been solved with Bioconductor/S4Vectors#46, which probably hasn't propagated to the build system yet. FYI, the reason that you get this error is described in passing in the `Note:` section of `?"Vector-comparison"` - _the more you know_...

> * It would be nice to have an option to drop the regions/levels that are not in use when calling `anchors(gi, type =1)` or `regions()`, analogous to dropping levels in a factor.

Probably better to have a separate `droplevels` method, rather than complicating the interface for `anchors` and `regions`. For example, someone sets the hypothetical `drop=TRUE` in `anchors`, does some operations on the indices, and tries to assign the indices back to the original GI object - but they don't make sense anymore, because the indices are all shifted when unused levels were dropped. Having `droplevels` explicitly create a new GI instance should reduce the surprise factor here.

Happy to take a PR here, though one would implement `droplevels,Factor-method` first.

> The class.Rmd vignette says that annotation should propagate to anchors retrieved with anchors(), but it doesn't.

Yes, this is a known problem. It's currently caused by how `GRangesFactor` objects `show` their "inner" `mcols` applied on the levels. For example:

```r
library(GenomicInteractions)
example(GenomicInteractions)
regions(test, 1)$something <- 1
first(test) # not shown...
levels(first(test)) # but it is there.
```

I had been thinking about this yesterday and it seems the situation is analogous to how the inner `mcols` are shown for `GRangesList` objects; if we consider the level `mcols` to be the inner `mcols` of a `GRangesFactor`, then arguably we should be showing them as well.

There is also a related issue of the default treatment of inner `mcols` during `unfactor` and `$`.

> * The `utilities.Rmd` vignette says that `pairdist(gi, type = "diagonal")` should work, but it only works with "diag".

Fixed to just use `"diag"`.

> I also found the description of this type a bit confusing, specifically, it's not clear to me what you mean by "this corresponds to a diagonal on the interaction space". 

I regret putting this option in. But yes, your description is better.

> Do you have a use case in mind that needs relative rather than absolute distances in bins?

With the benefit of hindsight, no. But if you're going for absolute distances, then `type="diag"` isn't the one you should be using. Arguably you could say that `type="diag"` is not useful in any circumstance, and I'd find it hard to disagree, but there's probably some code somewhere that uses it, so whatever.

**Comment by liz-is** [comment]
> which probably hasn't propagated to the build system yet. 

Ah, I guess it was just bad timing then!

> Probably better to have a separate droplevels method, rather than complicating the interface for anchors and regions.

I agree, it sounds like a `droplevels` method would make more sense.

> If we consider the level mcols to be the inner mcols of a GRangesFactor, then arguably we should be showing them as well.

Yes, I would agree with this. Does this imply there are also "outer" mcols, which might contain information that differs across different occurrences of a level? e.g., a column storing the annotation for the interacting region. Would these currently be shown instead of the "inner" mcols? 

Either way, I think some clarification in the vignette would be helpful.

> Arguably you could say that type="diag" is not useful in any circumstance, and I'd find it hard to disagree, but there's probably some code somewhere that uses it, so whatever.

Ha, fair enough!
